src/Api.js

```diff
@@ -89,13 +92,68 @@ class Api {
 	}
 
 	/**
-	 * Get the WhoColor data for a given wiki page.
+	 * Get the WikiWho data for a given wiki page. Use onError and onResults to set return handlers.
 	 *
-	 * @param  {string} wikiUrl URL of the wiki page that we want to analyze.
-	 * @return {Promise}
+	 * @param {string} wikiUrl URL of the wiki page that we want to analyze.
 	 */
 	getData( wikiUrl ) {
-		return $.getJSON( this.getAjaxURL( wikiUrl ) );
+		$.getJSON( this.getAjaxURL( wikiUrl ) ).then( ( function ( result ) {
+			// Handle error.
+			if ( !result.success && result.info && this.errorHandler ) {
+				// Turn the English error message into a i18n message key.
+				// @TODO Remove this when the WikiWho API returns useful error codes.
+				let errorCode = result.info
+					.toLowerCase()
+					.replace( / /g, '-' )
+					.replace( /[^a-z-]/g, '' );
+				// Possible messages:
+				// eslint-disable-next-line
+				// ext-whowrotethat-requested-data-is-not-currently-available-in-wikiwho-database-it-will-be-available-soon
+				this.errorHandler( 'ext-whowrotethat-' + errorCode );
+				return;
+			}
+
+			let cleanedResults = {
+				replacementHtml: result.extended_html,
+				users: {},
+				revisionsByUser: {}
+			};
+
+			// Reorganise the revisions to group by user ID.
+			// revisions structure: {rev_id: [timestamp, parent_rev_id, user_id, editor_name ], ...}
+			for ( let [ revId, rev ] of Object.entries( result.revisions ) ) {
+				let userId = rev[ 2 ];
+				if ( cleanedResults.revisionsByUser[ userId ] === undefined ) {
+					cleanedResults.revisionsByUser[ userId ] = {};
+				}
+				cleanedResults.revisionsByUser[ userId ][ revId ] = {
```

Comment:
(I'm not sure about this, checking) -- when we have a token, do we have userID or revID on it? Do we need the data structure organized by the user with all their revs, or by rev with whatever data (userid,etc) for easy fetching?

I suspect you're organizing correctly, but I also suspect we might want a way to check directly on a revision (?) if that's true, then even adding another map that holds a reference to the same data could help. I might be wrong here, though, or it might not be needed just yet?

Reply:
No, I was misunderstanding things; see the latest code for a better (maybe!) structure.